Cleanup & Fix main executable

v4hn opened the pull request:
I just tried this out for the first time and it didn't work.
Got it to work with these patches though.

I still can't convince myself it's a good idea to
couple the hardware simulation to the srdf, but anyway...

@davetcoleman @henningkayser @mlautman: Do you actually use this anywhere?
Do you directly use `ros_control_boilerplate sim_hw_main` or `fake_joint` instead?
We even have *partial* support for this in the setup assistant now, but it's all a bit of a mess though and no simulator actually supports velocities well...

v4hn committed:
if the pose does not exist, still initialize joints

Otherwise this warning should be fatal...

v4hn committed:
conditionally call loop only when available

The interface changed in melodic and now run() has to be called.

v4hn commented:
Apparently the repository has a broken CI config. The failure is unrelated.

JafarAbdi commented:
Closing and reopening for triggering Travis

mlautman commented:
> Closing and reopening for triggering Travis

In the future, you can re-trigger a test through the travis page rather than closing and opening a PR. If you don't have the necessary permissions to do that I would recommend talking to Dave and getting the correct permissions.

mlautman commented:
@v4hn I searched through three years of projects and found 2 that use this package. For one of those projects we were building `ros_control_boilerplate` from a source checkout of release `0.4.1`. 

This seems to enable us to build that project with the `melodic-devel` branch of `ros_control_boilerplate`.

mlautman commented:
> I still can't convince myself it's a good idea to
> couple the hardware simulation to the srdf, but anyway...

If I understand you correctly, you are referring to the joint model group requiring a loaded srdf to be able to figure out the joint model group to apply initial an initial joint pose. I don't really see the issue with this but maybe I am missing something. Either way, an alternative I guess would be to explicitly list both the joint names and joint values as rosparams.

mlautman commented:
Looks like `ROS_DISTRO=kinetic` is causing the failure because control_loop.run() isn't available in the kinetic branch of ros_control_boilerplate.

mlautman reviewed (approved):
You may want to run clang-format on this. Also, you may need to remove `ROS_DISTRO=kinetic` from `.travis.yaml` or add the melodic branch of ros_control_boilerplates to the rosinstall to get travis to pass.

v4hn commented:
> You may want to run clang-format on this

The project does not include a clang-format file and I get multiple unrelated changes if I use moveit's configuration. I adjusted whitespace in a comment though, if that's what you noticed.

> Looks like ROS_DISTRO=kinetic is causing the failure because control_loop.run() isn't available in the kinetic branch of ros_control_boilerplate.

[The branch actually includes `run()`](https://github.com/PickNikRobotics/ros_control_boilerplate/blob/kinetic-devel/src/generic_hw_control_loop.cpp#L67), but the latest *released* version for kinetic does not. @JafarAbdi I'm not sure you are still interested in kinetic releases, but this seems like neglect to me.
I will leave it to you whether you want to release boilerplate for kinetic, add it to the CI workspace here, or just drop kinetic compatibility from the `melodic-devel` branch.

v4hn commented:
> > I still can't convince myself it's a good idea to
> > couple the hardware simulation to the srdf, but anyway...
> 
> If I understand you correctly, you are referring to the joint model group requiring a loaded srdf to be able to figure out the joint model group to apply initial an initial joint pose. I don't really see the issue with this but maybe I am missing something.

Yes, my concerns are primarily for the use-case where you want to use the robot simulation *without* MoveIt, even if it's just meant for a tutorial where users should setup their own robot config based on a simulated robot.
As a replacement for the `fake_controller_manager` it is ok in theory to rely on the config, but if we have a ros_control based simulation, why restrict it to MoveIt?

> Either way, an alternative I guess would be to explicitly list both the joint names and joint values as rosparams.

Yes, but neither package allows for this option as far as I can see.
[fake_joint](https://github.com/tork-a/fake_joint) does, but it does not provide a velocity interface - though the one in these packages is also rather crude with discontinuous velocity profiles and broken position checks.

v4hn committed:
bump cmake version

Stop CMP0048 warning.
https://github.com/ros-planning/geometric_shapes/pull/129

v4hn commented:
I fixed the build for both APIs of `ros_control_boilerplate` and added a micro patch for the cmake version.

Notice though that this is actually only a problem with the latest *released* version of the package on kinetic. The `kinetic-devel` branch actually includes this unfortunate API change.

@mlautman please merge.

davetcoleman commented:
> Yes, my concerns are primarily for the use-case where you want to use the robot simulation without MoveIt, even if it's just meant for a tutorial where users should setup their own robot config based on a simulated robot.

I made a MoveIt-agnostic ros_control simulator called [SimHWInterface](https://github.com/PickNikRobotics/moveit_sim_controller/blob/melodic-devel/include/moveit_sim_controller/moveit_sim_hw_interface.h#L55).

However when robots startup in simulation, they needed initial joint states. Rather than making another standard for this, I just piggy backed on SRDF.

My thoughts on moveit_sim_controller is that its for MoveIt use cases only (hence the name), but builds on top of ros_control, which a better standard than the entirely MoveIt-specific fake_controller_manager.

v4hn commented:
> I made a MoveIt-agnostic ros_control simulator called SimHWInterface.

I know and I've used it in the past.

> However when robots startup in simulation, they needed initial joint states. Rather than making another standard for this, I just piggy backed on SRDF.

True and I somewhat changed my opinion about this since I wrote the comment.
It's still useful to have this node succeed startup without having a default group state available or specified. That's what this patch did.
